asciidoc/courses/app-go/modules/1-driver/lessons/1-installation/lesson.adoc

```diff
@@ -0,0 +1,166 @@
+----
+include::{repository-raw}/{branch}/pkg/example/example.go[tag=driver]
+----
+
+The function expects two arguments;
+
+1. A Connection URI - for example, `neo4j://localhost:7687` or `neo4j+s://dbhash.databases.neo4j.io`
+2. An authentication _token_
+// 3. Optionally, link:https://neo4j.com/docs/api/Go-driver/current/function/index.html#static-function-driver[additional driver configuration^].
+
+
+[TIP]
+.Authentication Types
+====
+Neo4j supports four authentication methods, basic authentication (e.g. username and password or LDAP), a base64 encoded Bearer token, a Kerberos token, or a custom authentication token.
+For more information on these options, visit link:https://neo4j.com/docs/operations-manual/current/authentication-authorization/[Authentication and Authorization^].
+====
+
+
+=== Verifying Connectivity
+
+You can verify that the connection details used during driver instantiation are correct by calling the `VerifyConnectivity()` method.
+
+This function returns a Promise which throws a `Neo.ClientError.Security.Unauthorized` error if the provided credentials are incorrect.
+
+.Verify Connectivity
+[source,go,indent=0,role=nocopy,subs="attributes+"]
+----
+include::{repository-raw}/{branch}/pkg/example/example.go[tag=verify]
+----
+
+== Sessions
+
+From the Driver object, you will open one or more **sessions**.
+A session is a client-side abstraction that borrows underlying TCP connections from a _connection pool_ held by the Driver and uses them to communicate with Neo4j.
+
+Sessions should be considered lightweight and disposable and can be opened and closed at will.
+
+You should `defer` the closing of each session  object using the `PanicOnClosureError()` function, which will release any underlying TCP connections held by the session object once the surrounding context has completed.
+
+This practise should also be followed when you are finished using the driver instance.
+
+.Session Lifecycle
+[source,go]
+----
+include::{repository-raw}/{branch}/pkg/example/example.go[tags=driver,close,indent=0]
+
+// Open a new Session
+session := driver.NewSession(ctx, neo4j.SessionConfig{})
+defer PanicOnClosureError(ctx, session)
+----
+
+
+== Running a Sample Query
+
+To run a one-off Cypher query, you can use the `session.Run()` method.
+
+.Running a One-off Query
+[source,go,indent=0,role=nocopy,subs="attributes+"]
+----
+include::{repository-raw}/{branch}/pkg/example/example.go[tag=oneoff]
+----
+
+The method accepts four arguments:
+
+1. The execution context.
+2. A string to represent a Cypher statement
+3. Optionally, a map representing query parameters (prefixed in Cypher with `$`)
+4. Also optional, a map containing transaction config.
+
+
+[WARNING]
+.For one-off queries only
+If there are any transient errors when running a query, the Driver will not attempt to retry a query when using `session.Run()`.
+For this reason, these should only be used for one-off queries and not in production.
```

Comment:
not sure if `CALL {} IN TRANSACTIONS` and `... USING PERIODIC COMMIT` are worth mentioning here? (they HAVE to run with `session.Run`, other execution methods do not work)

Reply:
That's interesting. I didn't know that.  I'm not sure whether this is out of the scope of the course